internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -263,19 +239,35 @@ func (f *Fetcher) FetchExplicitDestinationsData(
 	return destinations, statuses, nil
 }
 
+type PortReferenceKey struct {
+	resource.ReferenceKey
+	Port string
+}
+
 // FetchImplicitDestinationsData fetches all implicit destinations and adds them to existing destinations.
 // If the implicit destination is already in addToDestinations, it will be skipped.
 // todo (ishustava): this function will eventually need to fetch implicit destinations from the ImplicitDestinations resource instead.
-func (f *Fetcher) FetchImplicitDestinationsData(ctx context.Context, proxyID *pbresource.ID, addToDestinations []*intermediateTypes.Destination) ([]*intermediateTypes.Destination, error) {
+func (f *Fetcher) FetchImplicitDestinationsData(
+	ctx context.Context,
+	proxyID *pbresource.ID,
+	addToDestinations []*intermediateTypes.Destination,
+) ([]*intermediateTypes.Destination, error) {
 	// First, convert existing destinations to a map so we can de-dup.
-	destinations := make(map[resource.ReferenceKey]*intermediateTypes.Destination)
+	//
+	// This is keyed by the serviceID+port of the upstream, which is effectively
+	// the same as the id of the computed routes for the service.
+	destinations := make(map[PortReferenceKey]*intermediateTypes.Destination)
 	for _, d := range addToDestinations {
-		destinations[resource.NewReferenceKey(d.ServiceEndpoints.Resource.Id)] = d
+		prk := PortReferenceKey{
```

Comment:
I suspect this was a subtle bug before, having the explicit/implicit de-dupe map not include port.